src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -932,13 +910,22 @@ protected override void EndProcessing()
             {
                 // Load the source if they want to load from a file
                 if (String.Equals(ParameterSetName, "FromPath", StringComparison.OrdinalIgnoreCase) ||
-                    String.Equals(ParameterSetName, "FromLiteralPath", StringComparison.OrdinalIgnoreCase))
+                    String.Equals(ParameterSetName, "FromLiteralPath", StringComparison.OrdinalIgnoreCase)
+                   )
                 {
-                    this.sourceCode = "";
+                    StringBuilder sb = new StringBuilder(paths.Length);
+
                     foreach (string file in paths)
                     {
-                        this.sourceCode += System.IO.File.ReadAllText(file) + "\n";
+                        sb.Append(System.IO.File.ReadAllText(file));
```

Comment:
Some feedback:
1. `ReadAllText` returns a large string in the end, which may be allocated in LOH if the file size is larger than 85kb. The allocation of this large string will be wasted since we are appending it to the StringBuilder.
If using `ReadAllLines`, then only small strings will be created, which can be quickly garbage collected. So it probably would be better if we go with
```
foreach (string file in paths)
{
    foreach (string line in File.ReadAllLines(file))
    {
        sb.AppendLine(line);
    }
}
sourceCode = sb.ToString();
```
2. When there is only one file specified, which might be the case for the most time, this is still not performant. It could be better if we can treat the one-file case specially. Like:
```
if (paths.Count == 1)
{
    sourceCode = File.ReadAllText(file);
}
else
{
    foreach (string file in paths)
    {
        foreach (string line in File.ReadAllLines(file))
        {
            sb.AppendLine(line);
        }
    }
    sourceCode = sb.ToString();
}
```

Reply:
My thoughts was that we should get `CompileAssemblyFromFile` and there is no point in doing a lot of optimization.  But we can get it for a while, so I'll add the optimization. Only it will work for small files - `StringBuilder.MaxCapacity` is `int.MaxValue`.

Reply:
I don't understand your comment about `CompileAssemblyFromFile` (is it a class?). Could you please elaborate it a bit more? I quickly search about `CSharpSyntaxTree.ParseText`, and didn't find a good way to make it consume multiple files all together directly from files. That API takes a `SourceText` type, which has a static method `Create` that takes a file path, but not multiple files. But I'm sure there should be ways to do it, otherwise, how would a compiler to work 😄.

For some background, the perf issue of Add-Type was reported by Azure Profiler team. They find this exact code causes a lot LOH allocations in some services and if we treat the `path.length == 1` case specially, we can save, as quoted, "5.39 million samples here, or around 1% of total PowerShell.exe CPU cost". It was about the windows powershell, but this piece of code is same for powershell core.

Reply:
>didn't find a good way to make it consume multiple files all together directly from files

I didn't find too. So I decided to postpone this to a later PR.
Really `CompileAssemblyFromFile` is in CodeDom. There isn't such method In CodeAnalysis.
[I found](https://stackoverflow.com/questions/31175881/unable-to-get-method-syntaxtree-parsefile-in-new-nuget-of-roslyn) using `SourceText.From`.
Maybe @lzybkr know a right way.